### PR TITLE
True object and array streaming

### PR DESCRIPTION
**Pull request by csaboka**
As the comments say, `getArrayStream()` and `getObjectStream()` were very simplistic and didn't conform to the spirit of the specification, which requires lazy loading from the underlying stream.

Now these methods should behave lazily, only keeping the minimal amount of data in memory that's necessary for the stream.

### CODE REVIEW

File: johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java [hunk 1]
```diff
@@ -134,18 +140,57 @@ public void skipArray() {
...
-        //X TODO this implementation is very simplistic
-        //X I find it unintuitive what the spec intends here
-        //X we probably need to improve this
-        return getArray().stream();
+        Event current = current();
+        if (current != Event.START_ARRAY) {
+            throw new IllegalStateException(current + " doesn't support getArrayStream()");
+        }
+
+        Spliterator<JsonValue> arraySpliterator = new Spliterators.AbstractSpliterator<JsonValue>(
```
Comment (rmannibucau) [inline johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java:148]: What about making it an inner private class reused in both locations (private static class)?

File: johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java [hunk 2]
```diff
@@ -134,18 +140,57 @@ public void skipArray() {
...
+                Long.MAX_VALUE, Spliterator.IMMUTABLE | Spliterator.NONNULL | Spliterator.ORDERED) {
+
+            @Override
+            public boolean tryAdvance(Consumer<? super JsonValue> action) {
+                Event next = next();
+
+                if (next == Event.END_ARRAY) {
+                    return false;
+                } else {
```
Comment (rmannibucau) [inline johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java:157]: no need of else since first branch "returns"

**Review by rmannibucau — CHANGES_REQUESTED** [review]
2 small picky details to update to get it perfectly but overall looks  good
would be great to add error cases too in tests to ensures it behaves properly (getObjectStream does not test key event properly for example so error should be more precise than the one we would get right now IMHO)

File: johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java (hunk 1 above)
Comment (csaboka) [reply johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java:148]: The generic arguments and the `tryAdvance()` implementations are different, so what could be reused? As far as I can see, the only thing duplicated between the two Spliterator implementations are the arguments passed to the superclass constructor. I can extract that to a helper class if you'd like, but it feels overkill to me.

**Review by csaboka — COMMENTED** [review]
> would be great to add error cases too in tests to ensures it behaves properly (getObjectStream does not test key event properly for example so error should be more precise than the one we would get right now IMHO)

Can you clarify this a bit? Should I check against the result of the first `next()` call in `getObjectStream` to make sure it returned `KEY_NAME`? As far as I can tell there is no need to check the value of the second call because `getValue` will complain anyway if the parser is in the wrong state.

**Comment by rmannibucau** [comment]
@csaboka it is mainly about ensuring the error message references it is about the key and not just a "string" (which can be a value), think we should harness that

File: johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java (hunk 1 above)
Comment (rmannibucau) [reply johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java:148]: if you think you need 2 inner classes it is fine for me too, what I want to ensure is that it is not referencing $0 (this).
it is also fine ot inject the impl of tryAdvance as a lambda (the compiler will make it right normally) so maybe something along:

> new SimpleSplitIterator(action -> {  Event next = next(); ..... });

this enables to factorize the config and avoids the $0 ref.

Comment (csaboka) [reply johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java:148]: It needs to reference the enclosing parser instance, though, because it is calling methods on that instance like `next()`. If `tryAdvance` is factored out into a lambda, it will still capture `this`. As far as I can tell, a lambda won't help anything in this situation. It either needs to stay a nonstatic class like it is now, or it can be turned into a static class, but instances of that static class will need to explicitly reference the parser they came from.

Comment (rmannibucau) [reply johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java:148]: Pass parser as a param ;)
Lambdas have 3 conversions so if well refactored it solves it.
You can use javap to ensure you reached that goal

Comment (csaboka) [reply johnzon-core/src/main/java/org/apache/johnzon/core/JohnzonJsonParserImpl.java]: I've pushed some changes. Is this what you had in mind?

**Comment by csaboka** [comment]
> @csaboka it is mainly about ensuring the error message references it is about the key and not just a "string" (which can be a value), think we should harness that

I've added a check to make sure we get `KEY_NAME` in the object stream (unless it's `END_OBJECT` and we're done). I couldn't come up with a good test case for it, though. All the cases I could think of where `next()` would return something other than `KEY_NAME` ended up failing with a `JsonParsingException` instead.

**Comment by rmannibucau** [comment]
While consistent with parser api it is fine, just needs to be harnessed to ensure this consistency.
If you cant get a more precise message it is not critical for now but test should be there IMHO.

Side note: ill be off computer for some time so if somebody wants to merge once the test is added consider you got my +1

**Comment by csaboka** [comment]
Added some unit tests to validate that a parse exception is thrown when trying to traverse the stream and invalid JSON is encountered. I still can't figure out a good way to make the `KEY_NAME` check to fail. I could write a new subclass of JohnzonJsonParserImpl just to make it return unexpected values from `next()`, but that seems to be overkill for testing this edge case.

File: johnzon-core/src/test/java/org/apache/johnzon/core/JsonParserStreamingTest.java [hunk 3]
```diff
@@ -121,4 +127,59 @@ private String parserAndConcat(String json) {
...
+        expectedMappings.put("something", "else");
+        assertEquals(expectedMappings, mappings);
+    }
+
+    @Test(expected = JsonParsingException.class)
+    public void testParseErrorInGetObjectStream() {
+        StringReader sr = new StringReader("{42}");
+        JsonParser jsonParser = Json.createParser(sr);
+
+        JsonParser.Event firstEvent = jsonParser.next();
```
Comment (rmannibucau) [inline johnzon-core/src/test/java/org/apache/johnzon/core/JsonParserStreamingTest.java:179]: guess we should use a try with resource pattern to close the parser (it is required even for string readers due to memory management, even if not critical in tests it is always better to not show antipatterns in tests since they are kind of part of the doc)
if you test the message it looks ok.

But looks sufficient for me as coverage.

Comment (csaboka) [reply johnzon-core/src/test/java/org/apache/johnzon/core/JsonParserStreamingTest.java]: None of the other tests closed the parser, that's why I assumed it's fine. I've changed now so the parser is closed not just in the new tests, but the existing ones as well.

**Review by rmannibucau — APPROVED** [review]
